crates/ika-core/src/dwallet_mpc/cryptographic_computations_orchestrator.rs

```diff
@@ -1,130 +1,216 @@
+    Completed,
+}
+
+/// The orchestrator for DWallet MPC cryptographic computations.
+///
+/// The orchestrator's job is to manage a task queue for computations
+/// and avoid launching tasks that cannot be parallelized at the moment
+/// due to unavailable CPUs.
+/// When a CPU core is freed, and before launching the Rayon task,
+/// it ensures that the computation has not become redundant based on
+/// messages received since it was added to the queue. This approach
+/// reduces the read delay from the local DB without slowing down state sync.
+pub(crate) struct CryptographicComputationsOrchestrator {
+    /// The number of logical CPUs available for cryptographic computations on the validator's
+    /// machine.
+    available_cores_for_cryptographic_computations: usize,
+    /// A channel sender to notify the manager that a computation has been completed.
+    /// This is needed to decrease the [`currently_running_sessions_count`] when a computation is
+    /// done.
+    computation_channel_sender: UnboundedSender<ComputationUpdate>,
+    /// The number of currently running cryptographic computations — i.e.,
+    /// computations we called [`rayon::spawn_fifo`] for,
+    /// but we didn't receive a completion message for.
+    currently_running_sessions_count: usize,
+    epoch_store: Arc<AuthorityPerEpochStore>,
+}
+
+impl CryptographicComputationsOrchestrator {
+    /// Creates a new orchestrator for cryptographic computations.
+    pub(crate) fn try_new(epoch_store: &Arc<AuthorityPerEpochStore>) -> DwalletMPCResult<Self> {
+        let completed_computation_channel_sender =
+            Self::listen_for_completed_computations(&epoch_store);
+        let available_cores_for_computations: usize = std::thread::available_parallelism()
+            .map_err(|e| DwalletMPCError::FailedToGetAvailableParallelism(e.to_string()))?
+            .into();
+        if !(available_cores_for_computations > 0) {
+            return Err(DwalletMPCError::InsufficientCPUCores);
+        }
+
+        Ok(CryptographicComputationsOrchestrator {
+            available_cores_for_cryptographic_computations: available_cores_for_computations,
+            computation_channel_sender: completed_computation_channel_sender,
+            currently_running_sessions_count: 0,
+            epoch_store: epoch_store.clone(),
+        })
+    }
+
+    fn listen_for_completed_computations(
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) -> UnboundedSender<ComputationUpdate> {
+        let (completed_computation_channel_sender, mut completed_computation_channel_receiver) =
+            tokio::sync::mpsc::unbounded_channel();
+        let epoch_store_for_channel = epoch_store.clone();
+        tokio::spawn(async move {
+            loop {
+                match completed_computation_channel_receiver.recv().await {
+                    None => {
+                        break;
+                    }
+                    Some(updateValue) => match updateValue {
```

Comment:
```suggestion
                    Some(update_value) => match update_value {
```
This is not Typescript :)